### ISSUE TITLE
Support for services in childAdditions introduced in 10.14

### ISSUE THREAD

**Issue by mbay-ODW** [reubenmiller/go-c8y-cli#171]
![image](https://user-images.githubusercontent.com/58465677/185399823-4ff48b5d-0347-41cf-9d04-1267d2ad248d.png)
Since 10.14 events/alarms/measurements can directly be attached to an service. This service is an childAddition. However there is currently no easy way to directly create an e.g. event on that service.

As a workaround I can get the id via an additional request and use --device {id} which works, however an implementation via --service or something would be nice ;-)

**Comment by reubenmiller** [comment]
@mbay-ODW Can you provide more details on how the service is represented in c8y? Is the service a child (asset|addition|device) of the device? And on which managed object is the event/alarm/measurement being added to?

**Comment by mbay-ODW** [comment]
Happy to help @reubenmiller.

In deed it is a childAdditions:

![image](https://user-images.githubusercontent.com/58465677/185427609-93f8f3dc-904f-4bbc-9a6e-0e89e3cf1207.png)

The MO has the type: c8y_Service

![image](https://user-images.githubusercontent.com/58465677/185427836-ddc96b13-eafe-4284-9f99-ab575c4f1468.png)


The measurements/events/alarms are send to the childAdditions:

![image](https://user-images.githubusercontent.com/58465677/185428233-6b0308b6-d1ca-46d9-854a-c51b85f54db9.png)

**Comment by k-butz** [comment]
E.g. to create a measurement for a service named dockerd on a given device you could use: 
`c8y devices children list --id mytestdevice --childType addition --query "type eq c8y_Service and name eq dockerd" | c8y measurements create --data [...]`

But not sure if that's the most elegant solution. Let's see what @reubenmiller comes up with...

**Comment by reubenmiller** [comment]
**Warning: Only showing ideas, the examples below are not functioning examples!**

Yes the latest release (v2.15.2) at least makes it easier to filter child types, however it is not super user friendly as it requires writing long commands, and you don't get any command completion. However the solution is not straight forward as the service name still has the dependency on the device.

One proposal could be to add the `c8y devices services` subcommand:

```sh
# List all services
c8y devices services list --device "mydevices"

# Get a specific service by name
c8y devices services get --device "mydevices" --service dockerd

# Then to add any additional elements to the service, you just pipe the result to other commands
c8y devices services get --device "mydevices" --service dockerd | c8y measurements create --template "{...}"
```

Another approach that I have been thinking about for a while would be to add a new way to specify what kind of named lookup (provider) you are wanting to use to resolve the entity you are working with. For example:

The provider definition would follow the syntax of something like:

```sh
<provider>::<name>

# Or build child relationships using the -> operator
<provider>::<name> -> <provider>::<name_of_child>
```

```sh
# Lookup a device by name (use ~ "device::" prefix)
c8y events create --source "device::mydevicename"

# Build up child relationships by using ~ "->"
c8y measurements create --source "device::mydevice -> service::dockerd"

# Or something more generic for child relationships...
c8y measurements create --source "device::mydevice -> addition::dockerd"

# Additional "lookup" providers could be created to also support looking up via the external Id
c8y alarms create --source "identity::MyUniqueID"
c8y alarms create --source "identity::c8y_Serial:MyUniqueID"
```

One advantage with the `provider` pattern is that it would allow you to customize how you lookup agents/devices/inventory etc,  as usually different businesses have different ways of referring to their devices.

Though we have to be careful that we don't complicate this interface too much, and don't introduce too much "magic".

**Comment by mbay-ODW** [comment]
Looks already pretty great.

What I like on Proposal 1:

1. Clean
2. Stays in the old structure
3. Device Centric (From Device A give me Service Y)

What I like on Proposal 2:

1. Moving from --device to --source might come at any point anyway because measurements/alarms/events will at some point in time be written to assets and not just devices anymore
2. Very powerful structure/concept with the provider approach

I agree with you that Proposal 2 is more complicated, but does something speak against a coexistence of both approaches?

**Comment by reubenmiller** [comment]
Included in [v2.16.0](https://github.com/reubenmiller/go-c8y-cli/releases/tag/v2.16.0) and docs are available [here](https://goc8ycli.netlify.app/docs/cli/c8y/devices/services/c8y_devices_services/)

I went with option 1, though renamed some of the commands slightly.
